### PR TITLE
agent: Reduce startup time by 4 seconds

### PR DESCRIPTION
**Pull request by krishicks**
The reason `nomad agent` takes so long to start up is because of fingerprinting for AWS and GCE, each of which has a 2 second timeout by default, and because that code was run sequentially.

In the previous implementation, we would attempt to run each fingerprinter once, and then maybe create a goroutine to run the fingerprinter periodically if the fingerprinter was periodic.

Instead, we use that goroutine to run the initial fingerprinting, and either return early or enter a forever loop based on whether the fingerprinter is periodic or not.

This also required a small change to the debug logging. Previously, the initial list of fingerprinters would be given in a single log line:

```
...
    2020-11-18T09:13:13.009-0800 [DEBUG] client.fingerprint_mgr: detected fingerprints: node_attrs=[arch, bridge, cgroup, cpu, host, network, nomad, signal, storage]
...
```

But now they are printed individually as they are discovered:

```
...
    2020-11-18T09:14:54.328-0800 [DEBUG] client.fingerprint_mgr: detected fingerprint: name=cpu
    2020-11-18T09:14:54.329-0800 [DEBUG] client.fingerprint_mgr: detected fingerprint: name=host
    2020-11-18T09:14:54.329-0800 [INFO]  client.fingerprint_mgr.cgroup: cgroups are available
    2020-11-18T09:14:54.329-0800 [DEBUG] client.fingerprint_mgr: detected fingerprint: name=cgroup
...
```

### CODE REVIEW

**Comment by krishicks** [comment]
Oh, there's Azure also, so really, it's more like 6 seconds.

File: client/fingerprint_manager.go [hunk 1]
```diff
@@ -122,42 +122,40 @@ func (fm *FingerprintManager) Reload() {
...
-	fm.logger.Debug("detected fingerprints", "node_attrs", appliedFingerprints)
 	return nil
 }
 
 // runFingerprint runs each fingerprinter individually on an ongoing basis
-func (fm *FingerprintManager) runFingerprint(f fingerprint.Fingerprint, period time.Duration, name string) {
+func (fm *FingerprintManager) runFingerprint(f fingerprint.Fingerprint, name string) {
+	detected, err := fm.fingerprint(name, f)
+	if err != nil {
+		fm.logger.Debug("error periodic fingerprinting", "error", err, "fingerprinter", name)
```
Comment (cgbaker) [inline client/fingerprint_manager.go:146]:
```suggestion
		fm.logger.Debug("error fingerprinting", "error", err, "fingerprinter", name)
```

**Comment by jippi** [comment]
How will it affect restarting/reloading nomad during an in-place upgrade of a host - where jobs are scheduled with properties using these fingerprints? Wondering if Nomad will either stop/reschedule the job since the client doesn't have a fingerprint, or will temporarily create blocked eval

Example constraint

```
constraint {
    attribute = "${meta.aws.instance.availability-zone}"
    value     = "us-east-1d"
  }
```

File: client/fingerprint_manager.go [hunk 2]
```diff
@@ -122,42 +122,40 @@ func (fm *FingerprintManager) Reload() {
...
+		return
```
Comment (tgross) [inline client/fingerprint_manager.go:147]: The interesting thing I'm seeing here is that originally we would be returning exiting early for the entire `setupFingerprinters`, preventing us from running any other fingerprinters. That seems like we might be not just improving performance but fixing a bug 👍

**Review by tgross — APPROVED** [review]
LGTM 👍

**Review by cgbaker — APPROVED** [review]
👍

File: client/fingerprint_manager.go [hunk 3]
```diff
@@ -122,42 +122,40 @@ func (fm *FingerprintManager) Reload() {
...
-
 	for _, name := range fingerprints {
 		f, err := fingerprint.NewFingerprint(name, fm.logger)
-
 		if err != nil {
 			fm.logger.Error("error fingerprinting", "error", err, "fingerprinter", name)
 			return err
 		}
 
-		detected, err := fm.fingerprint(name, f)
```
Comment (jippi) [inline client/fingerprint_manager.go:135]: the underlying fingerprinting `err` here is no longer bubbling up in the new implementation, but only logged as a debug line - isn't that a behavior change?

Comment (krishicks) [reply client/fingerprint_manager.go:135]: It is a change in behavior. As @tgross noted above, previously the failure of any single fingerprinter was considered fatal. With this change that is no longer the case. I'll think on this some.

**Comment by notnoop** [comment]
Does this affect how the first fingerprint is sent to the client?  Would it be possible to send a fingerprint update to the server before AWS/GCE gets submitted?  If so, we may risk not running system jobs on newly created nodes if the constraints don't match: https://github.com/hashicorp/nomad/issues/6482

**Comment by jippi** [comment]
> Does this affect how the first fingerprint is sent to the client? Would it be possible to send a fingerprint update to the server before AWS/GCE gets submitted? If so, we may risk not running system jobs on newly created nodes if the constraints don't match: #6482

I raised similar in https://github.com/hashicorp/nomad/pull/9388#issuecomment-729828882

**Comment by krishicks** [comment]
I took a look at the code which is now running sooner than it would have before because of the updates in this PR.

It looks like it would drop down to [here](https://github.com/hashicorp/nomad/blob/hicks/fingerprint/client/client.go#L519), where it will wait for the `fpInitialized` channel to return, or the timeout (55 seconds) to elapse. `fpInitialized` is closed in `batchFirstFingerprints`, which is run in a goroutine not long after the fingerprinters are run: https://github.com/hashicorp/nomad/blob/hicks/fingerprint/client/client.go#L464

I don't know this code very well, so I'm not sure if this is enough information to assuage the above concerns or not. But it _seems_ like it's already configured to wait for the initial fingerprints to have occurred prior to sending off the initial batch.

**Comment by krishicks** [comment]
After looking into this further, I think we'll need to choose a different approach to get the startup time down. In addition to the above, there's a concern in the code that the GCE/EC2/Azure fingerprinters run last, as they may provide more detail than the prior fingerprinters. This change parallelizes the whole set, which breaks that contract.

Thanks to those that reviewed it!

**Pull request closed without merging**